src/server/mod.rs

```diff
@@ -336,6 +405,381 @@ impl<T, B> Service for HttpService<T>
+            inner.cancel()
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// The return type of `Http::bind_upgradable_connection`.
+///
+/// If the server task shuts down normally, this future completes with a `None`
+/// output value.
+///
+/// If an `Upgrade` response is returned by the bound `Service`, the server task
+/// is shut down (without closing the underlying I/O object) and the provided
+/// upgrade information is forwarded to this future, along with the underlying
+/// I/O object and any remaining buffered data that has already been read from
+/// it.
+///
+/// In the case of an error handling the connection, this future will complete
+/// with that error.
+pub struct BindUpgradableConnection<T, B, P> {
+    receiver: UpgradableReceiver<T, B, P>,
+    state: Option<(http::Conn<T, B, http::ServerTransaction>, P)>,
+}
+
+impl<T, B, P> fmt::Debug for BindUpgradableConnection<T, B, P>
+    where B: AsRef<[u8]>,
+          P: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("BindUpgradableConnection")
+            .field("receiver", &Omitted)
+            .field("state", &self.state)
+            .finish()
+    }
+}
+
+impl<T, B, P> Future for BindUpgradableConnection<T, B, P>
+    where T: AsyncRead + AsyncWrite + 'static,
+          B: AsRef<[u8]> + 'static,
+{
+    type Item = Option<(T, BytesMut, P)>;
+    type Error = ::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            if let Some((ref mut conn, _)) = self.state {
+                try_ready!(conn.poll_complete());
+            }
+
+            if let Some((conn, upgrade_info)) = self.state.take() {
+                // We ensure above that the connection is flushed before
+                // unwrapping.
+                let (io, read_buf) = conn.try_into_inner().ok().unwrap();
+                return Ok(Some((io, read_buf, upgrade_info)).into());
+            }
+
+            match self.receiver.poll() {
+                Err(_) => return Ok(None.into()),
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Ok(Async::Ready(result)) => {
+                    match result {
+                        Err(err) => return Err(err),
+                        Ok((transport, upgrade_info)) => {
+                            self.state = Some((transport.0, upgrade_info));
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Either an HTTP response or a signal to upgrade from HTTP to another
+/// protocol.
+///
+/// If the latter, the `Upgrade` variant should contain information pertaining
+/// to this protocol switch.
+pub enum UpgradableResponse<P = (), B = ::Body> {
+    /// An HTTP response.
+    Response(Response<B>),
+    /// A protocol upgrade signal with accompanying information and optional
+    /// "switching protocols" HTTP response head.
+    Upgrade(P, Option<Response<()>>),
```

Comment:
So, what exactly is the `P`? From what I can tell, is it just some arbitrary thing that might be created when inspected a `Request`, and useful once upgraded?